File: coolercontrol-gui/coolercontrol/services/liquidctl_device_extractors/corsair_hid_psu.py
#  CoolerControl - monitor and control your cooling and other devices
#  Copyright (c) 2022  Guy Boldon
#  |
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  |
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  |
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.
# ----------------------------------------------------------------------------------------------------------------------

import logging
from typing import List, Dict, Any

from liquidctl.driver.corsair_hid_psu import CorsairHidPsu

from coolercontrol.models.channel_info import ChannelInfo
from coolercontrol.models.device_info import DeviceInfo
from coolercontrol.models.lighting_mode import LightingMode
from coolercontrol.models.speed_options import SpeedOptions
from coolercontrol.models.status import TempStatus, ChannelStatus
from coolercontrol.services.liquidctl_device_extractors import LiquidctlDeviceInfoExtractor

_LOG = logging.getLogger(__name__)


# pylint: disable=protected-access
class CorsairHidPSUExtractor(LiquidctlDeviceInfoExtractor):
    supported_driver = CorsairHidPsu
    _channels: Dict[str, ChannelInfo] = {}
    _lighting_speeds: List[str] = []
    _min_temp = 20
    _max_temp = 100

    @classmethod
    def extract_info(cls, device_instance: CorsairHidPsu) -> DeviceInfo:
        cls._channels['fan'] = ChannelInfo(
            speed_options=SpeedOptions(
                min_duty=30,
                max_duty=100,
                profiles_enabled=False,
                fixed_enabled=True,
                manual_profiles_enabled=True
            )
        )

        return DeviceInfo(
            channels=cls._channels,
            lighting_speeds=cls._lighting_speeds,
            temp_min=cls._min_temp,
            temp_max=cls._max_temp,
            temp_ext_available=True
        )

    @classmethod
    def _get_filtered_color_channel_modes(cls, channel_name: str) -> List[LightingMode]:
        return []

    @classmethod
    def _get_temperatures(cls, status_dict: Dict[str, Any], device_id: int) -> List[TempStatus]:
        temps = []
        vrm_temp = cls._get_vrm_temp(status_dict)
        if vrm_temp is not None:
            temps.append(TempStatus("vrm", vrm_temp, "VRM", f"LC#{device_id} VRM"))
        case_temp = cls._get_case_temp(status_dict)
        if case_temp is not None:
            temps.append(TempStatus("case", case_temp, "Case", f"LC#{device_id} Case"))
        return temps

    @classmethod
    def _get_channel_statuses(cls, status_dict: Dict[str, Any]) -> List[ChannelStatus]:
        channel_statuses: List[ChannelStatus] = []
        fan_rpm = cls._get_fan_rpm(status_dict)
        if fan_rpm is not None:
            channel_statuses.append(ChannelStatus('fan', rpm=fan_rpm))
        return channel_statuses

File: coolercontrold/daemon/src/repositories/liquidctl/supported_devices/corsair_hid_psu.rs
/*
 * CoolerControl - monitor and control your cooling and other devices
 * Copyright (c) 2021-2025  Guy Boldon, Eren Simsek and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

use std::collections::HashMap;

use crate::device::{
    ChannelInfo, ChannelStatus, DeviceInfo, DriverInfo, DriverType, LightingMode, SpeedOptions,
};
use crate::repositories::liquidctl::base_driver::BaseDriver;
use crate::repositories::liquidctl::liqctld_client::DeviceResponse;
use crate::repositories::liquidctl::supported_devices::device_support::{DeviceSupport, StatusMap};

#[derive(Debug)]
pub struct CorsairHidPsuSupport;
// corsair_hid_psu.py

impl CorsairHidPsuSupport {
    pub fn new() -> Self {
        Self {}
    }
}

impl DeviceSupport for CorsairHidPsuSupport {
    fn supported_driver(&self) -> BaseDriver {
        BaseDriver::CorsairHidPsu
    }

    fn extract_info(&self, device_response: &DeviceResponse) -> DeviceInfo {
        let mut channels = HashMap::new();
        channels.insert(
            "fan".to_string(),
            ChannelInfo {
                speed_options: Some(SpeedOptions {
                    // We override the driver's min duty of 30%.
                    // 15% is half as fast rpm-wise, noticeably less noisy, and still offers
                    // decent cooling in lower power draw situations.
                    // Tested on real hardware and collaborates with other reports.
                    // The hardware has a different response when attempting to apply <13% and is
                    // pretty much limited to that as the slowest speed achievable for the fan
                    // itself. Manually setting 0 rpm is not possible, except in auto/hardware mode,
                    // and it takes a long time to kick down. Note that in auto mode 0 rpm can
                    // happen a lot, so it's often best to just use auto mode.
                    min_duty: 15,
                    max_duty: 100,
                    fixed_enabled: true,
                    extension: None,
                }),
                ..Default::default()
            },
        );
        channels.insert(
            "total-power".to_string(),
            ChannelInfo {
                label: Some("Total Power".to_owned()),
                ..Default::default()
            },
        );
        channels.insert(
            "estimated-input-power".to_string(),
            ChannelInfo {
                label: Some("Estimated Input Power".to_owned()),
                ..Default::default()
            },
        );
        channels.insert(
            "12v-power".to_string(),
            ChannelInfo {
                label: Some("+12V Power".to_owned()),
                ..Default::default()
            },
        );
        channels.insert(
            "5v-power".to_string(),
            ChannelInfo {
                label: Some("+5V Power".to_owned()),
                ..Default::default()
            },
        );
        channels.insert(
            "3.3v-power".to_string(),
            ChannelInfo {
                label: Some("+3.3V Power".to_owned()),
                ..Default::default()
            },
        );
        DeviceInfo {
            channels,
            lighting_speeds: Vec::new(),
            temp_min: 0, // device has vrm and case temps
            temp_max: 100,
            driver_info: DriverInfo {
                drv_type: DriverType::Liquidctl,
                name: Some(self.supported_driver().to_string()),
                version: device_response.liquidctl_version.clone(),
                locations: self.collect_driver_locations(device_response),
            },
            ..Default::default()
        }
    }

    fn get_color_channel_modes(&self, _channel_name: Option<&str>) -> Vec<LightingMode> {
        Vec::new()
    }

    fn get_channel_statuses(
        &self,
        status_map: &StatusMap,
        _device_index: u8,
    ) -> Vec<ChannelStatus> {
        let mut channel_statuses = vec![];
        self.add_single_fan_status(status_map, &mut channel_statuses);
        self.add_psu_power_statuses(status_map, &mut channel_statuses);
        channel_statuses.sort_unstable_by(|s1, s2| s1.name.cmp(&s2.name));
        channel_statuses
    }
}
